lib/mix/lib/mix/scm/git.ex

```diff
@@ -125,14 +125,18 @@ defmodule Mix.SCM.Git do
     sparse_toggle(opts)
     update_origin(opts[:git])
 
+    rev = get_lock_rev(opts[:lock], opts) || get_opts_rev(opts)
+
     # Fetch external data
     ["--git-dir=.git", "fetch", "--force", "--quiet"]
     |> Kernel.++(progress_switch(git_version()))
     |> Kernel.++(tags_switch(opts[:tag]))
+    |> Kernel.++(depth_switch(opts[:depth]))
+    |> Kernel.++(if rev, do: ["origin", rev], else: [])
```

Comment:
We could limit the explicit remote and rev for when `:depth` is used, otherwise there's a change in behavior here. Instead of fetching all branches etc from the remote, we'd fetch only what we're going to use next.

I think this is something to be discussed. Sticking to the old behavior is safest, but the new behavior can slightly improve performance.

Reply:
> we will be able to load other branches later?

Yes. Can freely fetch new branches/tags/refs and increase/decrease/remove depth.

The difference is whether or not it is expected that the clone contains "a reference all remote branches". I think that's an internal implementation detail on `mix`, but I can be wrong. To my defense, this change didn't break any existing test :smile: 

When I [proposed using `mix` to fetch non-Elixir/non-Erlang code](https://github.com/phoenixframework/phoenix/issues/5622#issuecomment-1791103706), I was not sure I was doing something supported or not, so I don't know where exactly the boundaries are.

My intuition is that if anyone is accessing the repo managed by mix inside of `deps/dep-name`, they can only expect to find what they asked for explicitly in their `mix.exs`, or they are relying on undocumented behavior and would be better served by making their requirement explicit (like explicitly using git to manage/fetch refs, or cloning the repo somewhere else independent of `mix`).

Reply:
What we didn't account for was SHA1 prefixes used in `:ref`!

https://github.com/elixir-lang/elixir/pull/13713 reverts back to fetching all refs so that short refs can be resolved.